File: packages/ts/protocol/src/revocation.ts
/**
 * Build signed v3 revocation lists.
 *
 * A {@link RevocationList} names `(sub, nonce, exp)` cap-cert tuples (and/or whole
 * `revokedSubjects`) revoked by an issuer's root identity. The list is
 * self-authenticating: it carries the issuer's Ed25519 signature over the canonical
 * serialization (`sig` stripped) plus a monotonic `generation` counter, so a server
 * can verify it without a cap and reject stale generations.
 *
 * This is the reusable builder the SDKs and apps were previously forced to hand-roll
 * (the example chat app signed lists inline). It mirrors the Python
 * `build_revocation_list` byte-for-byte — guarded by the shared
 * `tests/test-vectors/revocation-list.json` conformance vector.
 */
import { sha256 } from "@noble/hashes/sha2.js"
import { getBase64 } from "./platform.js"
import { stableStringify } from "./hash.js"
import * as ed25519Suite from "./suites/ed25519.js"
import { bytesToHex, hexToBytes } from "./suites/_hex.js"
import { userIdFromPubHex } from "./cap.js"

/** A single revoked cap-cert, identified by its subject + nonce, with the cap's expiry. */
export interface RevocationEntry {
  sub: string
  nonce: string
  exp: number
}

/** Revokes every cap with this subject (incident-response fallback). */
export interface RevokedSubject {
  sub: string
  exp: number
}

/** A signed, generation-counted revocation list issued by a root identity. */
export interface RevocationList {
  v: 1
  iss: string
  issUserId: string
  generation: number
  revoked: RevocationEntry[]
  revokedSubjects?: RevokedSubject[]
  /** Ed25519 signature over the canonical signing input, base64-encoded. */
  sig: string
}

export interface BuildRevocationListOpts {
  issEdPubHex: string
  issEdPrivHex: string
  generation: number
  revoked: RevocationEntry[]
  revokedSubjects?: RevokedSubject[]
}


/**
 * Domain-separation tag prepended to a revocation-list signing input. Binds the
 * signature to the "revocation-list" message type by construction so it can
 * never be reinterpreted as a cap-cert or request signature. Must stay
 * byte-identical across TS, Python, and the test-vector generators.
 */
const REVOCATION_DOMAIN = "starfish-revlist-v1\n"

/**
 * Canonical signing input for a revocation list: `stableStringify` of the list
 * with `sig` stripped. Byte-for-byte identical to the Python
 * `revocation_list_canonical_signing_input`.
 */
export function revocationListCanonicalSigningInput(
  list: Omit<RevocationList, "sig"> | RevocationList,
): string {
  const unsigned: Record<string, unknown> = { ...(list as Record<string, unknown>) }
  delete unsigned.sig
  return REVOCATION_DOMAIN + stableStringify(unsigned)
}

/**
 * Build and sign a {@link RevocationList}. `issUserId` is derived as
 * `sha256(issEdPub)[0:32]`; `revokedSubjects` is included (and signed) only when
 * supplied.
 */
export function buildRevocationList(opts: BuildRevocationListOpts): RevocationList {
  const unsigned: Omit<RevocationList, "sig"> = {
    v: 1,
    iss: opts.issEdPubHex,
    issUserId: userIdFromPubHex(opts.issEdPubHex),
    generation: opts.generation,
    revoked: opts.revoked,
    ...(opts.revokedSubjects !== undefined ? { revokedSubjects: opts.revokedSubjects } : {}),
  }
  const message = new TextEncoder().encode(revocationListCanonicalSigningInput(unsigned))
  const sigBytes = ed25519Suite.sign(message, opts.issEdPrivHex)
  return { ...unsigned, sig: getBase64().encode(sigBytes) }
}

File: packages/python/protocol/starfish_protocol/revocation.py
"""Build signed v3 revocation lists.

A ``RevocationList`` names ``(sub, nonce, exp)`` cap-cert tuples (and/or whole
``revokedSubjects``) revoked by an issuer's root identity. The list is
self-authenticating: it carries the issuer's Ed25519 signature over the
canonical serialization (``sig`` stripped) plus a monotonic ``generation``
counter, so a server can verify it without a cap and reject stale generations.
"""

from __future__ import annotations

import base64
from typing import Any

from starfish_protocol.cap import _user_id_from_pub_hex
from starfish_protocol.hash import stable_stringify
from starfish_protocol.suites import ed25519 as ed25519_suite


_REVOCATION_DOMAIN = "starfish-revlist-v1\n"


def revocation_list_canonical_signing_input(revocation_list: dict[str, Any]) -> str:
    """Canonical signing input for a revocation list."""
    unsigned = {k: v for k, v in revocation_list.items() if k != "sig"}
    return _REVOCATION_DOMAIN + stable_stringify(unsigned)


def build_revocation_list(
    iss_ed_pub_hex: str,
    iss_ed_priv_hex: str,
    generation: int,
    revoked: list[dict[str, Any]],
    revoked_subjects: list[dict[str, Any]] | None = None,
) -> dict[str, Any]:
    """Build and sign a ``RevocationList`` with the issuer's Ed25519 key."""
    unsigned: dict[str, Any] = {
        "v": 1,
        "iss": iss_ed_pub_hex,
        "issUserId": _user_id_from_pub_hex(iss_ed_pub_hex),
        "generation": generation,
        "revoked": revoked,
    }
    if revoked_subjects is not None:
        unsigned["revokedSubjects"] = revoked_subjects
    message = revocation_list_canonical_signing_input(unsigned).encode("utf-8")
    sig = base64.b64encode(ed25519_suite.sign(message, iss_ed_priv_hex)).decode("ascii")
    return {**unsigned, "sig": sig}
